### PR TITLE
Saving state of detailed coverage files in each cycle for each trial.

### PR DESCRIPTION
**Pull request by BharathMonash**
Changes have been made to the `detailed_coverage_data_utils.py` to incorporate the generation csv files in each cycle for each trial which will be copied to the GCS bucket later.

There is function which I think Jonathan has written and is planning to use it as call to save all the states. I have extended it to incorporate the functionality to record the state of the detailed coverage data files, although, as of now, Jonathan is not invoking that function anywhere currently, but upon invocation it will save the state of the detailed coverage files.

### CODE REVIEW

File: experiment/measurer/measure_manager.py [hunk 1]
```diff
@@ -606,6 +581,12 @@ def get_fuzzer_stats(self, cycle):
             logger.error('Stats are invalid.')
             return None
 
+    def save_state(self, cycle):
```
Comment (BharathMonash) [inline experiment/measurer/measure_manager.py:613]: This is the function which he wants to use as the top-level call.

File: experiment/measurer/measure_manager.py [hunk 2]
```diff
@@ -580,20 +522,53 @@ def process_crashes(self, cycle):
...
+        ones."""
+        previous_files = self.get_prev_measured_files(cycle,
+                                                      detailed_coverage=True)
+        current_segment_files = set(os.listdir(self.segment_dir))
+        current_function_files = set(os.listdir(self.function_dir))
+        all_files = previous_files.union(current_function_files,
+                                         current_segment_files)
+        measured_files_state = self.get_measured_files_state(cycle)
+        measured_files_state.set_current(list(all_files))
+        return all_files
```
Comment (mboehme) [inline experiment/measurer/measure_manager.py]: So, you are storing the segment and function file *names* in the state file? Why?

File: experiment/measurer/detailed_coverage_data_utils.py [hunk 3]
```diff
@@ -31,123 +31,68 @@ class DetailedCoverageData:  # pylint: disable=too-many-instance-attributes
...
-        csv_filestore_helper('segments.csv.gz', self.segment_df)
-        csv_filestore_helper('names.csv.gz', self.name_df)
+        csv_filestore_helper(
+            'functions_{trial_id}_{cycle}.csv.gz'.format(trial_id=trial_id,
+                                                         cycle=cycle),
+            self.function_df, 'function')
+        csv_filestore_helper(
+            'segments_{trial_id}_{cycle}.csv.gz'.format(trial_id=trial_id,
+                                                        cycle=cycle),
+            self.segment_df, 'segment')
```
Comment (mboehme) [inline experiment/measurer/detailed_coverage_data_utils.py]: All of this is being taken care of by the state file. Why are you still trying to persist trial and cycle specific information yourself?

**Review by mboehme — COMMENTED** [review]
Thanks @BharathMonash! I think, we are still using the state file wrongly. A state file takes care of persisting any kind of information. We don't need to write CSV files to disk. What you store in the state file are simply the names of the CSV files that you manually persisted.

Instead it is much easier. You load the previous state file, read the coverage information from *all* previous cycles, add to that the coverage information from the current cycle, and store the state file again. For debugging, you can always persist the "current" trial-specific CSV.

**Comment by BharathMonash** [comment]
closing this branch on account of new PR #22 .

**Pull request closed without merging**